index.js

```diff
@@ -4,15 +4,155 @@
   }else{
-    global.concurrify = f();
+    global.concurrify = f(global.sanctuaryTypeClasses, global.sanctuaryTypeIdentifiers);
   }
 
-}(/*istanbul ignore next*/(global || window || this), function(){
+}(/*istanbul ignore next*/(global || window || this), function(Z, type){
 
   'use strict';
 
-  return function(){};
+  var $alt = 'fantasy-land/alt';
+  var $ap = 'fantasy-land/ap';
+  var $map = 'fantasy-land/map';
+  var $of = 'fantasy-land/of';
+  var $zero = 'fantasy-land/zero';
+  var $$type = '@@type';
+
+  var ordinal = ['first', 'second', 'third', 'fourth', 'fifth'];
+
+  function isFunction(f){
+    return typeof f === 'function';
+  }
+
+  function isBinary(f){
+    return f.length >= 2;
+  }
+
+  function isApplicativeRepr(Repr){
+    try{
+      return Z.Applicative.test(Z.of(Repr));
+    }catch(_){
+      return false;
+    }
+  }
+
+  function invalidArgument(it, at, expected, actual){
+    throw new TypeError(
+      it
+      + ' expects its '
+      + ordinal[at]
+      + ' argument to '
+      + expected
+      + '\n  Actual: '
+      + Z.toString(actual)
+    );
+  }
+
+  function invalidContext(it, actual, an){
+    throw new TypeError(
+      it + ' was invoked outside the context of a ' + an + '. \n  Called on: ' + Z.toString(actual)
+    );
+  }
+
+  function last(xs){
+    return xs[xs.length - 1];
+  }
+
+  function getReprType(Repr){
+    return Repr[$$type] || Repr.name || 'Anonymous';
+  }
+
+  function concurrentPrepender(x, i, xs){
+    return i === xs.length - 1 ? ('Concurrent' + x) : x;
+  }
+
+  function computeType(type){
+    return type.split('/').map(concurrentPrepender).join('/');
+  }
+
+  //concurrify :: Applicative m
+  //           => (TypeRep m, () -> m a, (m a, m a) -> m a, (m a, m (a -> b)) -> m b)
+  //           -> Concurrently m
+  return function concurrify(Repr, zero, race, ap){
+
+    var INNERTYPE = getReprType(Repr);
+    var OUTERTYPE = computeType(INNERTYPE);
+
+    function Concurrently(sequential){
+      this.sequential = sequential;
+    }
+
+    function isInner(x){
+      return x instanceof Repr
+      || (Boolean(x) && x.constructor === Repr)
+      || type(x) === Repr[$$type];
+    }
+
+    function isOuter(x){
+      return x instanceof Concurrently
+      || (Boolean(x) && x.constructor === Concurrently)
+      || type(x) === OUTERTYPE;
+    }
+
+    function construct(x){
+      if(!isInner(x)) invalidArgument(OUTERTYPE, 0, 'be of type "' + INNERTYPE + '"', type(x));
+      return new Concurrently(x);
+    }
+
+    if(!isApplicativeRepr(Repr)) invalidArgument('concurrify', 0, 'represent an Applicative', Repr);
+
+    if(!isFunction(zero)) invalidArgument('concurrify', 1, 'be a function', zero);
+    var _zero = zero();
+    if(!isInner(_zero)) invalidArgument('concurrify', 1, 'to return ' + INNERTYPE, type(_zero));
+
+    if(!isFunction(race)) invalidArgument('concurrify', 2, 'be a function', race);
+    if(!isBinary(race)) invalidArgument('concurrify', 2, 'be binary', race);
```

Comment:
The `isBinary` check might be annoying. In implementations of `alt` where the second argument is ignored it may force someone to wrap their implementation. This reinforces my thought that I may have been somewhat overzealous checking the input to `concurrify`.